### PR TITLE
Update Particle Container to Pure SoA

### PR DESCRIPTION
**Pull request by ax3l**
Transition particle containers to pure SoA layouts.

- [x] test on CUDA with MPI
- [x] test on OMP with MPI
- [x] wait for 1st and 2nd CPU particle `Redistribute` fix in AMReX
   - https://github.com/AMReX-Codes/amrex/pull/3296
   - https://github.com/AMReX-Codes/amrex/pull/3299
- [x] more AMReX updates:
  - https://github.com/AMReX-Codes/amrex/pull/3303
  - https://github.com/AMReX-Codes/amrex/pull/3290
- [x] merge WarpX first https://github.com/ECP-WarpX/WarpX/pull/3850
- [x] merge pyAMReX first https://github.com/AMReX-Codes/pyamrex/pull/124
- [x] improve performance of id handling via constants and https://github.com/AMReX-Codes/amrex/pull/3735
- [x] update docs to make sure AoS is not mentioned anymore and particle id handling is guided

## Benchmarks on NVIDIA A100-SXM 80GB GPU, SP

```
./impactx ../../examples/fodo/input_fodo.in beam.npart=600000000 diag.enable=false diag.slice_step_diagnostics=false lattice.elements=drift1 quad1 drift2 quad2 drift3
```
[soa_gpu_sp.txt](https://github.com/ECP-WarpX/impactx/files/14216314/soa_gpu_sp.txt), [dev_gpu_sp.txt](https://github.com/ECP-WarpX/impactx/files/14216522/dev_gpu_sp.txt)

- FODO runtime: **2.52x faster** :rocket: :sparkles: 
- `impactx::Push::Drift`: 2.33x faster
- `impactx::Push::Quad`: 2.85x faster
- `impactX::collect_lost_particles`: 2.77x faster
- `ImpactX::add_particles`: 5.4% faster

## Benchmarks on NVIDIA A100-SXM 80GB GPU, DP

```
./impactx ../../examples/fodo/input_fodo.in beam.npart=600000000 diag.enable=false diag.slice_step_diagnostics=false lattice.elements=drift1 quad1 drift2 quad2 drift3
```
[soa_gpu_dp.txt](https://github.com/ECP-WarpX/impactx/files/14216280/soa_gpu_dp.txt), [dev_gpu_dp.txt](https://github.com/ECP-WarpX/impactx/files/14216281/dev_gpu_dp.txt)

- FODO runtime: **1.73x faster** :rocket: :sparkles: 
- `impactx::Push::Drift`: 1.85x faster
- `impactx::Push::Quad`: 1.42x faster
- `impactX::collect_lost_particles`: 3.68x faster
- `ImpactX::add_particles`: 6.6% faster

## Benchmarks on CPU, SP
Laptop: 12th Gen Intel(R) Core(TM) i9-12900H, performance core w/ performance power mode on
```
export OMP_NUM_THREADS=1
./impactx ../../examples/fodo/input_fodo.in beam.npart=1000000 diag.enable=false diag.slice_step_diagnostics=false lattice.elements=drift1 quad1 drift2 quad2 drift3 &
taskset -cp 6 $!
```
[soa_cpu_sp.txt](https://github.com/ECP-WarpX/impactx/files/14216553/soa_cpu_sp.txt), [dev_cpu_sp.txt](https://github.com/ECP-WarpX/impactx/files/14216552/dev_cpu_sp.txt)

- FODO runtime: **1.28x faster** :rocket: :sparkles: 
- `impactx::Push::Drift`: 2.17x faster
- `impactx::Push::Quad`: 4.8% faster
- `impactX::collect_lost_particles`: 2.48x faster
- `ImpactX::add_particles`: 1.7% faster

## Benchmarks on CPU, DP
Laptop: 12th Gen Intel(R) Core(TM) i9-12900H, performance core w/ performance power mode on
```
export OMP_NUM_THREADS=1
./impactx ../../examples/fodo/input_fodo.in beam.npart=1000000 diag.enable=false diag.slice_step_diagnostics=false lattice.elements=drift1 quad1 drift2 quad2 drift3 &
taskset -cp 6 $!
```
[soa_cpu_dp.txt](https://github.com/ECP-WarpX/impactx/files/14216550/soa_cpu_dp.txt), [dev_cpu_dp.txt](https://github.com/ECP-WarpX/impactx/files/14216549/dev_cpu_dp.txt)

- FODO runtime: **1.21x faster** :rocket: :sparkles: 
- `impactx::Push::Drift`: 1.76x faster
- `impactx::Push::Quad`: same perf.
- `impactX::collect_lost_particles`: 3.24x faster
- `ImpactX::add_particles`: 2.2% faster

### CODE REVIEW

File: src/particles/elements/diagnostics/openPMD.cpp [hunk 1]
```diff
@@ -394,25 +365,36 @@ namespace detail
 
             for (auto real_idx=0; real_idx < RealSoA::nattribs; real_idx++) {
                 auto const component_name = real_soa_names.at(real_idx);
+                amrex::Print() << real_idx << " " << component_name << "\n";
```
Comment (ax3l) [inline src/particles/elements/diagnostics/openPMD.cpp]:
```suggestion
```

File: src/particles/ImpactXParticleContainer.cpp [hunk 2]
```diff
@@ -145,6 +135,7 @@ namespace impactx
          */
         auto old_np = particle_tile.numParticles();
         auto new_np = old_np + pinned_tile.numParticles();
+        amrex::AllPrint() << "old_np=" << old_np << " new_np=" << new_np << "\n";
```
Comment (ax3l) [inline src/particles/ImpactXParticleContainer.cpp]:
```suggestion
```

File: cmake/dependencies/ABLASTR.cmake [hunk 3]
```diff
@@ -163,18 +163,18 @@ set(ImpactX_openpmd_src ""
...
+set(ImpactX_ablastr_branch "topic-particle-soa"
     CACHE STRING
     "Repository branch for ImpactX_ablastr_repo if(ImpactX_ablastr_internal)")
 
 # AMReX is transitively pulled through ABLASTR
 set(ImpactX_amrex_repo "https://github.com/AMReX-Codes/amrex.git"
     CACHE STRING
     "Repository URI to pull and build AMReX from if(ImpactX_amrex_internal)")
-set(ImpactX_amrex_branch ""
+set(ImpactX_amrex_branch "a393d7ff7e320cefeeb55e31f1b0e0b5ac2d90ca"
```
Comment (ax3l) [inline cmake/dependencies/ABLASTR.cmake]:
```suggestion
set(ImpactX_amrex_branch "more_if_for_soa"
```

File: cmake/dependencies/ABLASTR.cmake [hunk 4]
```diff
@@ -163,18 +163,18 @@ set(ImpactX_openpmd_src ""
...
+set(ImpactX_ablastr_repo "https://github.com/ax3l/WarpX.git"
     CACHE STRING
     "Repository URI to pull and build ABLASTR from if(ImpactX_ablastr_internal)")
-set(ImpactX_ablastr_branch "23.05"
+set(ImpactX_ablastr_branch "topic-particle-soa"
     CACHE STRING
     "Repository branch for ImpactX_ablastr_repo if(ImpactX_ablastr_internal)")
 
 # AMReX is transitively pulled through ABLASTR
 set(ImpactX_amrex_repo "https://github.com/AMReX-Codes/amrex.git"
```
Comment (ax3l) [inline cmake/dependencies/ABLASTR.cmake]:
```suggestion
set(ImpactX_amrex_repo "https://github.com/atmyers/amrex.git"
```

**Comment by ax3l** [comment]
@cemitch99 when you have time next week, can you please review the chances I did to individual elements?

**Comment by ax3l** [comment]
@atmyers I am running on CUDA now after our latest fixes. It looks like most tests are doing well, but e.g., 1 MPI rank test `cfchannel_spacecharge.run` / `cfchannel_spacecharge.py.run` does not find its beam after this PR is applied (inits zero particles).

Same for:
- `expanding_beam.run`
- `kurth_10nC_periodic.run`
- `fodo_rf_sc.py.run`
- `bithermal` tests

Looks related to space charge.

There are more failing when running in `Debug` mode.

**Comment by cemitch99** [comment]
This pattern of changes to individual elements looks correct, and I like the improved symmetry between the treatment of position and momentum variables.  The main risk introduced is possible mixing of the initial values and final values, so I will need to check every element line-by-line before review.

**Comment by ax3l** [comment]
@atmyers some of the CUDA tests w/ space charge seem to fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  No particles found. Cannot run evolve without a beam.
```

So our init, resize and/or redistribute have an issue with this PR.

**Comment by ax3l** [comment]
Fixed :tada:

**Comment by ax3l** [comment]
Congrats to all who helped @Thierry992, @atmyers, @AlexanderSinn et al. 🎉 👏

**Comment by ax3l** [comment]
I repeated the benchmarks on GPU for scrutiny on Perlmutter A100 SMX 80 GB GPUs using the new AMReX https://github.com/AMReX-Codes/amrex/pull/3763 `tiny_profiler.device_synchronize_around_region = 1` syncs.

There is now a bit more overhead added due to the excessive syncs, but individual elements should be more precisely measured.

## Benchmarks on NVIDIA A100-SXM 80GB GPU, SP

```
./impactx ../../examples/fodo/input_fodo.in beam.npart=600000000 diag.enable=false diag.slice_step_diagnostics=false tiny_profiler.device_synchronize_around_region=1 lattice.elements=drift1 quad1 drift2 quad2 drift3
```
[soa_gpu_sp.txt](https://github.com/ECP-WarpX/impactx/files/14314773/soa_gpu_sp.txt), [old_gpu_sp.txt](https://github.com/ECP-WarpX/impactx/files/14315054/old_gpu_sp.txt)


- FODO runtime: **2.34x faster** :rocket: :sparkles: 
- `impactx::Push::Drift`: 2.32x faster
- `impactx::Push::Quad`: 2.32x faster
- `impactX::collect_lost_particles`: 2.78x faster
- `ImpactX::add_particles`: 5.2% faster

## Benchmarks on NVIDIA A100-SXM 80GB GPU, DP

```
./impactx ../../examples/fodo/input_fodo.in beam.npart=600000000 diag.enable=false diag.slice_step_diagnostics=false tiny_profiler.device_synchronize_around_region=1 lattice.elements=drift1 quad1 drift2 quad2 drift3
```
[soa_gpu_dp.txt](https://github.com/ECP-WarpX/impactx/files/14314775/soa_gpu_dp.txt), [old_gpu_dp.txt](https://github.com/ECP-WarpX/impactx/files/14315053/old_gpu_dp.txt)

- FODO runtime: **1.65x faster** :rocket: :sparkles: 
- `impactx::Push::Drift`: 1.80x faster
- `impactx::Push::Quad`: 1.32x faster
- `impactX::collect_lost_particles`: 3.65x faster
- `ImpactX::add_particles`: 1.6% faster

**Pull request merged**